### PR TITLE
fix omemo key loading from s2s servers

### PR DESCRIPTION
**Pull request by mightymop**
If a omemo session in a chatbox is initiated by the user, the keys and devicelist are now loaded from a pep service on another server (server 2 server connection) too.

### CODE REVIEW

File: src/plugins/omemo/Plugin.ts [hunk 1]
```diff
@@ -100,6 +103,42 @@ export default class OMEMOPlugin extends EncryptionPlugin {
       });
    }
 
+   private getDeviceListFromContact(contact: IContact)
+   {
+      let connection = this.pluginAPI.getConnection();
+      let _self = this;
+      return connection.getPEPService().retrieveItems('eu.siacs.conversations.axolotl.devicelist',contact.getJid().bare).then(function(data) {
+
+            return new Promise(function(resolve, reject) {
```
Comment (sualko) [inline src/plugins/omemo/Plugin.ts:112]: No need for `_self` with arrow functions.

File: src/plugins/omemo/Plugin.ts [hunk 2]
```diff
@@ -100,6 +103,42 @@ export default class OMEMOPlugin extends EncryptionPlugin {
...
+                if (_self.onDeviceListUpdate(data))
+                {
+                    const promises = [];
+                    let arr = $(data).find("device");
+                    let i=0;
+                    for (; i < arr.length; i++) {
+                       promises.push(connection.getPEPService().retrieveItems('eu.siacs.conversations.axolotl.bundles:'+$(arr[i]).attr('id'),contact.getJid().bare).then(function(data) {
```
Comment (sualko) [inline src/plugins/omemo/Plugin.ts:119]: No need to query all bundles. If a contact has a device list this should be enough.

File: src/plugins/omemo/Plugin.ts [hunk 3]
```diff
@@ -100,6 +103,42 @@ export default class OMEMOPlugin extends EncryptionPlugin {
...
+                            return new Promise(function(resolve, reject) {
```
Comment (sualko) [inline src/plugins/omemo/Plugin.ts:120]: No need to return a new promise.

File: src/plugins/omemo/Plugin.ts [hunk 4]
```diff
@@ -100,6 +103,42 @@ export default class OMEMOPlugin extends EncryptionPlugin {
...
+                       .catch((e) => {
+                           console.log(e);
+                           reject();
+                    });
+                    resolve(true);
+                }
+                else
+                {
+                    reject();
+                }
```
Comment (sualko) [inline src/plugins/omemo/Plugin.ts:137]: Should be simplified.

**Review by sualko — CHANGES_REQUESTED** [review]
Looks like a workaround for non working pep notifications. In normal circumstances, those chances should not be needed.

**Comment by mightymop** [comment]
> 
> 
> Looks like a workaround for non working pep notifications. In normal circumstances, those chances should not be needed.

You are right that normally it should not be needed... but i often get those errors like "Your contact does not support Omemo only because the keys where not received...". But before throwing such an error i like to check again to be sure. Especially on mobile devices when there were network changes outside and stream management is not available to detect package loss

File: src/plugins/omemo/Plugin.ts (hunk 2 above)
Comment (mightymop) [reply src/plugins/omemo/Plugin.ts:119]: I tried withouth receiving the bundles only the key, then i got this error that my contact does not support omemo.
Then i notice in the omemo device/key dialog that the keys are not loaded.
If i force to load the keys, then they are avaiable...

File: src/plugins/omemo/Plugin.ts (hunk 4 above)
Comment (mightymop) [reply src/plugins/omemo/Plugin.ts]: i added some comments..

File: src/plugins/omemo/Plugin.ts (hunk 3 above)
Comment (mightymop) [reply src/plugins/omemo/Plugin.ts]: changed

File: src/plugins/omemo/Plugin.ts (hunk 1 above)
Comment (mightymop) [reply src/plugins/omemo/Plugin.ts]: changed to lamda notation and removed _self var

**Comment by sualko** [comment]
> But before throwing such an error i like to check again to be sure

That's probably a good idea, but it should be enough to check the availability of the device list. The rest should be handled by the plugin. Otherwise the bundles would be requested twice. Also please use the already provided namespace constants and so on.

> if the contact is member of another server / domain, will pep notifications delivered to federated servers?

Usually this should work, but if you use OpenFire I can't give any guaranties.

**Comment by mightymop** [comment]
ok i removed the code for bundlereceiving...
maybe we should also load the devicelist if we open the omemo dialog... 
i opened before and there were no bundles until i enabled the lock icon and opened the dialog again...

**Comment by sualko** [comment]
> maybe we should also load the devicelist if we open the omemo dialog...

I think that's a good idea.

**Review by sualko — APPROVED** [review]

**Comment by sualko** [comment]
You `getDeviceListFromContact` always returned true and therefore every contact "supported" omemo :joy: 

I removed also the duplicated code in the dialog, but it should work now.

**Comment by mightymop** [comment]
> You `getDeviceListFromContact` always returned true and therefore every contact "supported" omemo 
> 
> I removed also the duplicated code in the dialog, but it should work now.

Hoppla... thats why we review code  😇 but i am increasing skills hopefully...

**Pull request merged**